pkg/filesystem/mount_unix.go

```diff
@@ -0,0 +1,130 @@
+//go:build darwin || freebsd || linux
+// +build darwin freebsd linux
+
+package filesystem
+
+import (
+	"errors"
+	"fmt"
+	"unsafe"
+
+	"github.com/buildbarn/bb-storage/pkg/util"
+
+	"golang.org/x/sys/unix"
+)
+
+const (
+	FSCONFIG_SET_FLAG        = 0
+	FSCONFIG_SET_STRING      = 1
+	FSCONFIG_SET_BINARY      = 2
+	FSCONFIG_SET_PATH        = 3
+	FSCONFIG_SET_PATH_EMPTY  = 4
+	FSCONFIG_SET_FD          = 5
+	FSCONFIG_CMD_CREATE      = 6
+	FSCONFIG_CMD_RECONFIGURE = 7
+)
+
+// Rudimentary function wrapper for the `fsconfig` syscall.
+//
+// This is implemented as a stop gap solution until real support is merged into the `unix` library.
+// See this patchset: https://go-review.googlesource.com/c/sys/+/399995/ .
+// This only implements the two commands needed for basic `mountat` functionality.
+// And will just exit if any other command is called.
+//
+// TODO(nils): construct proper `syscall.E*` errors.
+//	like `unix.errnoErr`, but the function is not exported.
+func fsconfig(fsfd int, cmd int, key string, value string, flags int) (err error) {
+	switch cmd {
+	case FSCONFIG_SET_STRING:
+		if len(key) == 0 || len(value) == 0 {
+			err = errors.New("`key` and `value` must be provided")
+			return
+		}
+	case FSCONFIG_CMD_CREATE:
+		if len(key) != 0 || len(value) != 0 {
+			err = errors.New("`key` and `value` must be empty")
+			return
+		}
+	default:
+		err = errors.New("not implemented: " + fmt.Sprintf("%d", cmd))
+		return
+	}
+
+	var _p0 *byte
+	var _p1 *byte
+
+	_p0, err = unix.BytePtrFromString(key)
+	if err != nil {
+		return
+	}
+	if key == "" {
+		_p0 = nil
+	}
+
+	_p1, err = unix.BytePtrFromString(value)
+	if err != nil {
+		return
+	}
+	if value == "" {
+		_p1 = nil
+	}
+
+	r0, _, e1 := unix.Syscall6(
```

Comment:
The same holds for `fsconfig()`. We could also let `golang.org/x/sys/unix` ship with a binding for that. My experience is that the folks maintaining that package are willing to accept such changes. They also review them in a timely manner.

Reply:
I have updated the commit with the patch file, and will wait a couple more days for them to see my comment on the original review before forking it.

Reply:
Pushed a new patch, with the new proposed API.

Reply:
There has not been any progress on the review since we last discussed the patch :/
https://github.com/golang/go/issues/59537#issuecomment-1752217786